src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -167,29 +209,25 @@ private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int leng
             // Get back Metrics[].
             if (!value2metrics.TryGetValue(tagValues, out aggregatorIndex))
             {
-                aggregatorIndex = this.metricPointIndex;
-                if (aggregatorIndex >= MaxMetricPoints)
-                {
-                    // sorry! out of data points.
-                    // TODO: Once we support cleanup of
-                    // unused points (typically with delta)
-                    // we can re-claim them here.
-                    return -1;
-                }
-
                 lock (value2metrics)
                 {
                     // check again after acquiring lock.
                     if (!value2metrics.TryGetValue(tagValues, out aggregatorIndex))
                     {
-                        aggregatorIndex = Interlocked.Increment(ref this.metricPointIndex);
-                        if (aggregatorIndex >= MaxMetricPoints)
+                        if (!this.metricPointFreeList.TryPop(out aggregatorIndex))
                         {
-                            // sorry! out of data points.
-                            // TODO: Once we support cleanup of
-                            // unused points (typically with delta)
-                            // we can re-claim them here.
-                            return -1;
+                            if (this.temporality == AggregationTemporality.Cumulative)
+                            {
+                                // sorry! out of data points.
+                                return -1;
+                            }
+
+                            this.FreeUnusedMetrics();
```

Comment:
since this call is done *inside* a lock and in hot path, could FreeUnusedMetrics() return as soon as it frees a single point, enough to unblock this measurement?